src/main/java/org/jabref/logic/integrity/DoiDuplicationChecker.java

```diff
@@ -0,0 +1,51 @@
+package org.jabref.logic.integrity;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import javafx.collections.ObservableList;
+
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.database.BibDatabase;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.StandardField;
+import org.jabref.model.entry.identifier.DOI;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+public class DoiDuplicationChecker implements Checker {
+    private final BibDatabase database;
+    private Map<BibEntry, List<IntegrityMessage>> errors;
+
+    public DoiDuplicationChecker(BibDatabase database) {
+        this.database = Objects.requireNonNull(database);
+    }
+
+    @Override
+    public List<IntegrityMessage> check(BibEntry entry) {
+        if (errors == null) {
+            errors = new HashMap<>();
```

Comment:
Why not directly initialize that HashMap in the field declaration?

Reply:
MAybe you did not see that the later code uses the BibtexDatabase from the **constructor**. We decided to put the code into check - and not into the constructor - to have a lean constructor and not have high CPU usage when initializing the checker, but at the first time an entry is checked.

Reply:
This seems like premature optimization for me.

Reply:
Why are you caching the errors anyway? Also in case you are worrying about performance, why do you choose to implement a solution that has O(n^2) (with n = entries). Why not implement a O(n) solution (e.g. https://stackoverflow.com/a/31341963/873661) which you call in `checkDatabase` below?

Reply:
See in line 30 that the method is called **once** per entry. Is is **by design** of the `Checker` interface that there is method for checking each entry. There is **no method** for checking a complete database. There is also no interface for a complete database.

At the first call, the result map is filled. At all subsequent calls do **not** fill the map and reuse it.

We are O(n+m) where n is the size of the database and m is the number of entries with DOIs.

Reply:
To make the code more reable, I moved the call to the error check to the constructor.